pkg/controller/common/helpers.go

```diff
@@ -1001,3 +1008,144 @@ func GetDefaultBaseImageContainer(cconfigspec *mcfgv1.ControllerConfigSpec) stri
+		return "", fmt.Errorf("%w: error running %s %s: %s", err, oc, strings.Join(args, " "), string(rawOut))
+	}
+	return strings.TrimSpace(string(rawOut)), nil
+}
+
+func generateICSPFile(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) (string, error) {
+	regisConf := &sysregistriesv2.V2RegistriesConf{}
+	// process the icspRules to registries.conf, it can merge mutiple icsps exist in icspRules
+	runtimeutils.EditRegistriesConfig(regisConf, nil, nil, icspRules)
+	icsp := apioperatorsv1alpha1.ImageContentSourcePolicy{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: apioperatorsv1alpha1.SchemeGroupVersion.String(),
+			Kind:       "ImageContentSourcePolicy",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "image-policy",
+		},
+	}
+	rdms := []apioperatorsv1alpha1.RepositoryDigestMirrors{}
+	for _, reg := range regisConf.Registries {
+		mirrors := []string{}
+		for _, m := range reg.Mirrors {
+			mirrors = append(mirrors, m.Location)
+		}
+		rdm := apioperatorsv1alpha1.RepositoryDigestMirrors{
+			Source:  reg.Location,
+			Mirrors: mirrors,
+		}
+		rdms = append(rdms, rdm)
+	}
+	icsp.Spec.RepositoryDigestMirrors = rdms
+
+	// Convert to json first so json tags are handled
+	jsonData, err := json.Marshal(&icsp)
+	if err != nil {
+		return "", err
+	}
+	contents, err := yaml.JSONToYAML(jsonData)
+	if err != nil {
+		return "", err
+	}
+	icspfile, err := os.CreateTemp("", "icsp-file")
+	if err != nil {
+		return "", err
+	}
+	if _, err := icspfile.Write(contents); err != nil {
+		icspfile.Close()
+		os.Remove(icspfile.Name())
+		return "", err
+	}
+	defer icspfile.Close()
+	return icspfile.Name(), nil
+}
+
+// ocImageInfoUseProxy uses oc image info to retrive the proxyTestImage, through the http(s) proxy if exists.
+// returns error if the command fails.
+func ocImageInfoUseProxy(oc, imageSpec, httpProxy, httpsProxy, icspfile string) error {
+	const (
+		// proxy error from docker registry starts with "proxyconnect"
+		proxyErr             = "proxyconnect"
+		cnoimageInfo         = "image info %s"
+		cnoimageInfoWithICSP = "image info %s --icsp-file %s"
+	)
+	var envs []string
+	if httpProxy != "" {
+		envs = append(envs, fmt.Sprintf("HTTP_PROXY=%s", httpProxy))
+	}
+	if httpsProxy != "" {
+		envs = append(envs, fmt.Sprintf("HTTPS_PROXY=%s", httpsProxy))
+	}
+
+	cmd := fmt.Sprintf(cnoimageInfo, imageSpec)
+	if icspfile != "" {
+		cmd = fmt.Sprintf(cnoimageInfoWithICSP, imageSpec, icspfile)
+	}
+	args := strings.Split(cmd, " ")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	cmdExec := exec.CommandContext(ctx, oc, args...)
+	cmdExec.Env = envs
+	rawOut, err := cmdExec.CombinedOutput()
+	if err != nil {
+		if strings.Contains(string(rawOut), proxyErr) {
+			return fmt.Errorf("invalid http proxy: %w: error running %s %s: %s", err, oc, strings.Join(args, " "), string(rawOut))
+		}
+		return fmt.Errorf("%w: error running %s %s: %s", err, oc, strings.Join(args, " "), string(rawOut))
```

Comment:
Should we have some kind of retry here, for transient failures?

I guess we always retry via re-syncing technically, maybe it would be worth adding a requeue somewhere...

The issue I'm thinking about is, let's say the network is unstable, and we happen to fail the one validation, but the proxy is otherwise valid, what is the user experience there?

Reply:
yes, I agree. we can retry to deal with the risks.